### ISSUE TITLE
Abnormal termination of GCHP operation caused by increasing the number of cores

### ISSUE THREAD

**Issue by helpyuan** [geoschem/gchp#340]
### Name and Institution (Required)

Name: liumy
Institution: AMS

### Confirm you have reviewed the following documentation

- [x] [Support guidelines](https://gchp.readthedocs.io/en/latest/reference/SUPPORT.html)
- [x] [User manuals](https://gchp.readthedocs.io/en/latest/geos-chem-shared-docs/supplemental-guides/related-docs.html)
- [x] [Debugging GEOS-Chem and HEMCO errors](https://gchp.readthedocs.io/en/latest/geos-chem-shared-docs/supplemental-guides/debug-guide.html)
- [x] [Current and past Github issues](https://github.com/geoschem/GCHP/issues)

### Description of your issue or question
The supercomputer I am using has 28 cores per thread, and I use it to run GCHP to simulate TransoprtTracers. But I encountered abnormal termination in the following different situations:
I can successfully run the simulation for one month using a total of 168 cores on 6 nodes; I can also use 9 nodes with a total of 252 cores to successfully run the simulation for one month. However, when I increased the number of nodes to 24 or 12, totaling 672 or 336 cores, GCHP experienced the following abnormal termination:

**1. 24 nodes , 672 cores:**
[1693466168.079281] [c884:28151:0]           sock.c:88   UCX  ERROR socket create failed: Too many open files
attachment
[672cores.err.txt](https://github.com/geoschem/GCHP/files/12483505/672cores.err.txt)
[672cores.out.txt](https://github.com/geoschem/GCHP/files/12483508/672cores.out.txt)
[allPEs_672.log.txt](https://github.com/geoschem/GCHP/files/12483509/allPEs_672.log.txt)
[core_672.log.txt](https://github.com/geoschem/GCHP/files/12483511/core_672.log.txt)
[input_672.nml.txt](https://github.com/geoschem/GCHP/files/12483512/input_672.nml.txt)
[logfile_672.000000.out.txt](https://github.com/geoschem/GCHP/files/12483513/logfile_672.000000.out.txt)
[setCommonRunSettings_672.sh.txt](https://github.com/geoschem/GCHP/files/12483514/setCommonRunSettings_672.sh.txt)

**2. 12 nodes, 336 cores:**
[1693466767.152946] [c891:22897:0]           sock.c:344  UCX  ERROR recv(fd=133) failed: Connection reset by peer
attachment
[336.err.txt](https://github.com/geoschem/GCHP/files/12483815/336.err.txt)
[336.out.txt](https://github.com/geoschem/GCHP/files/12483816/336.out.txt)
[allPEs_336.log](https://github.com/geoschem/GCHP/files/12483817/allPEs_336.log)
[core_336.log.txt](https://github.com/geoschem/GCHP/files/12483818/core_336.log.txt)
[input_336.nml.txt](https://github.com/geoschem/GCHP/files/12483819/input_336.nml.txt)
[logfile_336.000000.out.txt](https://github.com/geoschem/GCHP/files/12483820/logfile_336.000000.out.txt)
[PET029.ESMF_LogFile.txt](https://github.com/geoschem/GCHP/files/12483821/PET029.ESMF_LogFile.txt)
[setCommonRunSettings_336.sh.txt](https://github.com/geoschem/GCHP/files/12483822/setCommonRunSettings_336.sh.txt)

How can I solve this problem?

Please provide as much detail as possible. Always include the GCHP version number and any relevant configuration and log files.
GCHP/14.1.0
GCC/10.2.0
OPENMPI/4.1.0
ESMF/8,1.0

**Comment by lizziel** [comment]
Hi @helpyuan, for the first issue (too many open files) you may be having a system-specific issue due to your default settings. The problem is happening around reading the restart file. Take a look at this StackOverflow post and try some of the solutions. https://stackoverflow.com/questions/880557/socket-accept-too-many-open-files.

For your second issue you have an ESMF error (log file starting with PET). The run got past reading the restart file and the problem is during input regridding. Do you consistently get this error every time you run with that configuration?

Pinging @tclune and @mathomp4 in case they have seen this error before during ESMF regridding. The configuration is 12 nodes with 336 cores (28 cores per node). 

**ESMF error:** 
```
20230831 152607.123 ERROR            PET029 ESMCI_Mesh_Regrid_Glue.C:546 ESMCI_regrid_create() Internal error: Bad condition  - /data/home/liumy/tools/ESMF_8_1_0/esmf-8.1.0/src/Infrastructure/Mesh/src/Legacy/ESMCI_MeshDB.C, line:591:Bad meshobjtype in get_map = UNKNOWN
20230831 152607.123 ERROR            PET029 ESMCI_MeshCap.C:624 MeshCap::regrid_create() Internal error: Bad condition  - Internal subroutine call returned Error
20230831 152607.123 ERROR            PET029 ESMF_Regrid.F90:372 ESMF_RegridStore Internal error: Bad condition  - Internal subroutine call returned Error
20230831 152607.123 ERROR            PET029 ESMF_FieldRegrid.F90:1340 ESMF_FieldRegridStoreNX Internal error: Bad condition  - Internal subroutine call returned Error
```
**Run error:** 
```
sock.c:344 UCX ERROR recv(fd=133) failed: Connection reset by peer
attachment
```

**Comment by tclune** [comment]
Unfortunately, I've not seen this before.    Hopefully this implies it is some simple, albeit subtle, problem in the run configuration.

**Comment by mathomp4** [comment]
I'm going to ping @bena-nasa. If anyone has seen this, it might be him.

**Comment by lizziel** [comment]
@helpyuan, also try a run with ExtData debug prints on. For instruction on doing this see the Debugging guide on ReadTheDocs [here](https://gchp.readthedocs.io/en/latest/user-guide/debugging.html). More information on the `logging.yml` file is [here](https://gchp.readthedocs.io/en/14.1.0/user-guide/config-files/logging_yml.html). The output prints will go to log file `allPEs.log`. This may tell us what file is being regridded when the error is encountered.

I also suggest trying 24 cores per node to see if that makes a difference. Are you using mass flux input? There is a known dependency between resolution and number of cores when doing mass flux regridding.

**Comment by helpyuan** [comment]
Thanks for your replying @lizziel ! I will try a run with ExtData debug prints on.
Now, with your help, I have solved the first problem. But after running GCHP again, the same problem occurred as the second issue.

**"For your second issue you have an ESMF error (log file starting with PET). The run got past reading the restart file and the problem is during input regridding. Do you consistently get this error every time you run with that configuration?"**
Reply: Yes, this error occurs every time when I run with that configuration.

**"I also suggest trying 24 cores per node to see if that makes a difference. Are you using mass flux input? There is a known dependency between resolution and number of cores when doing mass flux regridding."**
Reply: I tried using 24 cores per node, but still encountered the same issue. 
`[1693652094.985401] [c808:13231:0]           sock.c:344  UCX  ERROR recv(fd=137) failed: Connection reset by peer
[1693652094.985309] [c802:14701:0]           sock.c:344  UCX  ERROR recv(fd=140) failed: Connection reset by peer`
Because I am running TransportTracers for the first time, I currently do not know what mass flux input is.  The following are my data input files for your reference.
[HEMCO_Config.rc.txt](https://github.com/geoschem/GCHP/files/12504126/HEMCO_Config.rc.txt)
[ExtData.rc.txt](https://github.com/geoschem/GCHP/files/12504127/ExtData.rc.txt)

**Comment by lizziel** [comment]
Hi @helpyuan, I'm glad you fixed the first problem. Does this mean you can successfully run with 24 nodes, but not with 12 nodes? Do you also see the error if you run at reduced grid resolution, e.g. C48?

You are not using input mass fluxes so that is not the issue. Using mass fluxes instead of winds (for advection) is an option when you create a run directory and choose GEOS-FP meteorology. It is independent of simulation type. You would see it listed in your ExtData.rc file as variables `MFXC` and `MFYC`.

**Comment by helpyuan** [comment]
Hi @lizziel, I'm sorry for replying to you so late.

**"Does this mean you can successfully run with 24 nodes, but not with 12 nodes?"**
I'm sorry, it's not. That means the running results with 24 nodes are the same as those with 12 nodes, still will terminate abnormally.

**Do you also see the error if you run at reduced grid resolution, e.g. C48?**
I cannot run c48 with 24 nodes because grid resolution is too low for core count requested, as shown follow: 
`ERROR: NX and NY are set such that face side length divided by NX or NY/6 is less than 4. The cubed sphere compute domain has a minimum requirement of 4 points in NX and NY/6. This problem occurs when grid resolution is too low for core count requested. Edit setCommonRunSettings.sh to loower total number of cores or increase your grid resolution.`
So I used 12 nodes to run c48 and set `CAP. EXTDATA` in `logging. yml` to "DEBUG".

When I run c48 with 12 nodes (28 cores per node), I still have the same error. The following are the relevant files. Due to the information in the PET**.log file is similar, I only uploaded PET000 and PET335. Due to transmission restrictions, I only uploaded the last 250000 lines in `allPEs.log`. If you need more information, please feel free to contact me.
[gchp.20190101_0000z.log.txt](https://github.com/geoschem/GCHP/files/12682609/gchp.20190101_0000z.log.txt)
[job.2123273.err.txt](https://github.com/geoschem/GCHP/files/12682613/job.2123273.err.txt)
[logfile.000000.out.txt](https://github.com/geoschem/GCHP/files/12682614/logfile.000000.out.txt)
[PET000.ESMF_LogFile.txt](https://github.com/geoschem/GCHP/files/12682615/PET000.ESMF_LogFile.txt)
[PET335.ESMF_LogFile.txt](https://github.com/geoschem/GCHP/files/12682616/PET335.ESMF_LogFile.txt)
[allPEs_lastlines_250000.log.txt](https://github.com/geoschem/GCHP/files/12682755/allPEs_lastlines_250000.log.txt)

In addition, when I was running 1 node (24 cores), the PET**.log file also showed information similar to that expressed by 12 nodes:
`20230921 105138.976 ERROR PET00 ESMF_Time.F90:815 ESMF_TimeGet() Object Set or SetDefault method not called - Object not Initialized`
But GCHP was still able to successfully run and generate Species concentration files. And in the last line of PET**.log, the similar infomation like `20230921 110853.297 INFO PET00 Finalizing ESMF` appears. I'm not sure if this is the problem caused by my settings, could you please help me answer it. Some relevant files are as follow:a
[gchp.20190101_0000z.log.txt](https://github.com/geoschem/GCHP/files/12682758/gchp.20190101_0000z.log.txt)
[job.2123118.err.txt](https://github.com/geoschem/GCHP/files/12682760/job.2123118.err.txt)
[logfile.000000.out.txt](https://github.com/geoschem/GCHP/files/12682762/logfile.000000.out.txt)
[PET00.ESMF_LogFile.txt](https://github.com/geoschem/GCHP/files/12682764/PET00.ESMF_LogFile.txt)

Finally, please let me know how to use more nodes (such as 12 nodes, 24 nodes, etc.) to successfully run GCHP. I would greatly appreciate it!

**Comment by lizziel** [comment]
Hi @helpyuan, are you able to get a C48 run to complete on two nodes? I wonder if there is an issue with your system preventing multi-node runs.

It is odd you are getting an ESMF error with a single node run and the run still completes. Try a 1-hour duration C24 run on a single node with the following settings in setCommonRunSettings.sh:

```
# Auto-update settings in HISTORY.rc for specific collections (enable with ON)
AutoUpdate_Diagnostics=ON

# Instructions to auto-update diagnostics
#   1. Set AutoUpdate_Diagnostics=ON:
#   2. Set Diag_Monthly to compute monthly time-averaged values (0=OFF, 1=ON)
#   3. If Diag_Monthly=OFF:
#        3a. Set Diag_Frequency for diagnostic frequency, format "HHmmSS"
#        3b. Set Diag_Duration for file write frequency, format "HHmmSS"
#        *Note that number of hours may exceed 2 digits, e.g. 744 for 744 hrs
#   4. Edit Diag_Collections list to specify which collections to update
#
Diag_Monthly="0"
Diag_Frequency="010000"
Diag_Duration="010000"
```

This will turn monthly diagnostics off. Monthly diagnostics can sometimes trip an ESMF error message without crashing the model, although I have not seen the specific error message you are getting.

**Comment by helpyuan** [comment]
Hi @lizziel , I have solved the issue when I updated ESMF-8.1.0 to ESMF-8.4.2. Thanks for your patient answer!

**Comment by lizziel** [comment]
Excellent! I will add information on ReadTheDocs about ESMF version. Perhaps it is time to up the minimum version requirement.
